### PR TITLE
Move tokenizers to new `olmo_data` package.

### LINKED ISSUE #633: Tokenizer with relative path import fails when using olmo as pip library
### 🐛 Describe the bug

I'm attempting to load a tokenizer from a saved checkpoint as follows, using OLMo as a library. I first pip-installed via:
```shell
pip install . [e]
```

Next, I tried to run the following code in a different repo:

```python
# Checkpoint downloaded from an OLMo-medium run
ckpt = "/net/nfs.cirrascale/allennlp/willm/olmo-sparsity/checkpoints/OLMo-medium/pile-llamaish7/step0"
tokenizer = Tokenizer.from_checkpoint(ckpt)
```

This results in the following error when the OLMo codebase tries to load a tokenizer from HuggingFace:

```markdown
Repository Not Found for url: https://huggingface.co/tokenizers/allenai_gpt-neox-olmo-dolma-v1_5.json/resolve/main/tokenizer.json.
Please make sure you specified the correct `repo_id` and `repo_type`.
If you are trying to access a private or gated repo, make sure you are authenticated.
```

It's unexpected that the tokenizer tries to load from HuggingFace because the `TokenizerConfig` that gets loaded from `ckpt` actually contains a relative path to a tokenizer in the OLMo repo:
```python
TokenizerConfig(identifier='tokenizers/allenai_gpt-neox-olmo-dolma-v1_5.json', truncate_direction='right')
```

As discussed with @dirkgr, the problem seems to be that, because I'm running my code from *my own repo* rather than within the OLMo repo, the relative path to the tokenizer fails.

This is undesired behavior when using OLMo as a pip package IMO. The user shouldn't be required to have a local clone of the repo since it is already installed somewhere via pip! Requiring the user to have a local clone is not just redundant, but also more concerningly could cause confusing behavior if the version of the pip installed OLMo and local OLMo are different.

### Versions

Python 3.10.9

<details>
<summary>pip freeze</summary>
<verbatim>
-e git+https://github.com/allenai/OLMo@d72a262645d831cc80d4a974718598998103075f#egg=ai2_olmo
 <br />-e git+https://github.com/allenai/OLMo-core@cda6f23421848b53f830fd98294ec49f68fc8637#egg=ai2_olmo_core
 <br />aiohttp==3.9.5
 <br />aiosignal==1.3.1
 <br />alabaster==0.7.16
 <br />annotated-types==0.6.0
 <br />antlr4-python3-runtime==4.9.3
 <br />anyio==3.7.0
 <br />argon2-cffi==21.3.0
 <br />argon2-cffi-bindings==21.2.0
 <br />arrow==1.2.3
 <br />asttokens==2.2.1
 <br />async-timeout==4.0.3
 <br />attrs==23.1.0
 <br />Babel==2.15.0
 <br />backcall==0.2.0
 <br />backoff==2.1.2
 <br />backports.tarfile==1.2.0
 <br />beaker-gantry==1.1.0
 <br />beaker-py==1.26.14
 <br />beautifulsoup4==4.12.2
 <br />black==23.12.1
 <br />bleach==6.0.0
 <br />blessed==1.20.0
 <br />blinker==1.8.2
 <br />boto3==1.34.96
 <br />boto3-extensions==0.23.0
 <br />botocore==1.34.96
 <br />brotlipy==0.7.0
 <br />build==1.2.1
 <br />cached_path==1.6.2
 <br />cachetools==5.3.3
 <br />certifi @ file:///croot/certifi_1671487769961/work/certifi
 <br />cffi @ file:///croot/cffi_1670423208954/work
 <br />charset-normalizer @ file:///tmp/build/80754af9/charset-normalizer_1630003229654/work
 <br />click==8.1.7
 <br />click-aliases==1.0.4
 <br />click-help-colors==0.9.1
 <br />cloudpickle==3.0.0
 <br />cmake==3.29.2
 <br />colorama==0.4.6
 <br />comm==0.1.3
 <br />conda==23.1.0
 <br />conda-content-trust @ file:///tmp/abs_5952f1c8-355c-4855-ad2e-538535021ba5h26t22e5/croots/recipe/conda-content-trust_1658126371814/work
 <br />conda-package-handling @ file:///croot/conda-package-handling_1672865015732/work
 <br />conda_package_streaming @ file:///croot/conda-package-streaming_1670508151586/work
 <br />contourpy==1.2.1
 <br />cryptography @ file:///croot/cryptography_1673298753778/work
 <br />cycler==0.12.1
 <br />datasets==2.7.1
 <br />dateparser==1.2.0
 <br />debugpy==1.6.7
 <br />decorator==5.1.1
 <br />defusedxml==0.7.1
 <br />dill==0.3.6
 <br />diskcache==5.6.3
 <br />distro==1.9.0
 <br />docker==6.1.3
 <br />docutils==0.20.1
 <br />evaluate==0.4.2
 <br />exceptiongroup==1.1.1
 <br />execnet==2.1.1
 <br />executing==1.2.0
 <br />fastapi==0.110.3
 <br />fastjsonschema==2.17.1
 <br />filelock==3.13.4
 <br />Flask==3.0.3
 <br />fonttools==4.51.0
 <br />fqdn==1.5.1
 <br />frozenlist==1.4.1
 <br />fsspec==2024.3.1
 <br />furo==2023.5.20
 <br />gitdb==4.0.10
 <br />GitPython==3.1.31
 <br />google-api-core==2.19.0
 <br />google-auth==2.30.0
 <br />google-cloud-core==2.4.1
 <br />google-cloud-storage==2.17.0
 <br />google-crc32c==1.5.0
 <br />google-resumable-media==2.7.1
 <br />googleapis-common-protos==1.63.1
 <br />gpustat==1.1
 <br />h11==0.14.0
 <br />halo==0.0.31
 <br />httpcore==1.0.5
 <br />httptools==0.6.1
 <br />httpx==0.27.0
 <br />huggingface-hub==0.21.4
 <br />idna @ file:///croot/idna_1666125576474/work
 <br />imagesize==1.4.1
 <br />importlib_metadata==7.2.0
 <br />iniconfig==2.0.0
 <br />interegular==0.3.3
 <br />ipykernel==6.23.2
 <br />ipython==8.14.0
 <br />ipython-genutils==0.2.0
 <br />ipywidgets==8.0.6
 <br />isodate==0.6.1
 <br />isoduration==20.11.0
 <br />isort==5.12.0
 <br />itsdangerous==2.2.0
 <br />jaraco.classes==3.4.0
 <br />jaraco.context==5.3.0
 <br />jaraco.functools==4.0.1
 <br />jedi==0.18.2
 <br />jeepney==0.8.0
 <br />Jinja2==3.1.2
 <br />jmespath==1.0.1
 <br />joblib==1.4.0
 <br />jsonpointer==2.3
 <br />jsonschema==4.17.3
 <br />jupyter==1.0.0
 <br />jupyter-console==6.6.3
 <br />jupyter-events==0.6.3
 <br />jupyter_client==8.2.0
 <br />jupyter_core==5.3.1
 <br />jupyter_server==2.6.0
 <br />jupyter_server_terminals==0.4.4
 <br />jupyterlab-pygments==0.2.2
 <br />jupyterlab-widgets==3.0.7
 <br />keyring==25.2.1
 <br />kiwisolver==1.4.5
 <br />lark==1.1.9
 <br />livereload==2.6.3
 <br />llvmlite==0.42.0
 <br />lm-format-enforcer==0.9.8
 <br />log-symbols==0.0.14
 <br />markdown-it-py==3.0.0
 <br />MarkupSafe==2.1.3
 <br />matplotlib==3.8.4
 <br />matplotlib-inline==0.1.6
 <br />maturin==1.5.1
 <br />mdit-py-plugins==0.4.1
 <br />mdurl==0.1.2
 <br />mistune==2.0.5
 <br />more-itertools==10.3.0
 <br />mpmath==1.3.0
 <br />msgpack==1.0.8
 <br />multidict==6.0.5
 <br />multiprocess==0.70.14
 <br />mypy==1.3.0
 <br />mypy-extensions==1.0.0
 <br />myst-parser==2.0.0
 <br />nbclassic==1.0.0
 <br />nbclient==0.8.0
 <br />nbconvert==7.5.0
 <br />nbformat==5.9.0
 <br />nest-asyncio==1.5.6
 <br />networkx==3.3
 <br />nh3==0.2.17
 <br />ninja==1.11.1.1
 <br />notebook==6.5.4
 <br />notebook_shim==0.2.3
 <br />numba==0.59.1
 <br />numpy==1.26.4
 <br />nvidia-cublas-cu12==12.1.3.1
 <br />nvidia-cuda-cupti-cu12==12.1.105
 <br />nvidia-cuda-nvrtc-cu12==12.1.105
 <br />nvidia-cuda-runtime-cu12==12.1.105
 <br />nvidia-cudnn-cu12==8.9.2.26
 <br />nvidia-cufft-cu12==11.0.2.54
 <br />nvidia-curand-cu12==10.3.2.106
 <br />nvidia-cusolver-cu12==11.4.5.107
 <br />nvidia-cusparse-cu12==12.1.0.106
 <br />nvidia-ml-py==11.525.112
 <br />nvidia-nccl-cu12==2.19.3
 <br />nvidia-nvjitlink-cu12==12.4.127
 <br />nvidia-nvtx-cu12==12.1.105
 <br />omegaconf==2.3.0
 <br />oocmap==0.3
 <br />openai==1.29.0
 <br />outlines==0.0.34
 <br />overrides==7.3.1
 <br />packaging==23.1
 <br />pandas==2.2.2
 <br />pandocfilters==1.5.0
 <br />parso==0.8.3
 <br />pathspec==0.12.1
 <br />petname==2.6
 <br />pexpect==4.8.0
 <br />pickleshare==0.7.5
 <br />pillow==10.3.0
 <br />pkginfo==1.11.1
 <br />platformdirs==3.5.3
 <br />pluggy==1.5.0
 <br />prometheus-fastapi-instrumentator==7.0.0
 <br />prometheus_client==0.20.0
 <br />prompt-toolkit==3.0.38
 <br />proto-plus==1.23.0
 <br />protobuf==4.25.3
 <br />psutil==5.9.5
 <br />ptyprocess==0.7.0
 <br />pure-eval==0.2.2
 <br />py-cpuinfo==9.0.0
 <br />pyarrow==16.0.0
 <br />pyasn1==0.6.0
 <br />pyasn1_modules==0.4.0
 <br />pycosat @ file:///croot/pycosat_1666805502580/work
 <br />pycparser @ file:///tmp/build/80754af9/pycparser_1636541352034/work
 <br />pydantic==2.7.1
 <br />pydantic_core==2.18.2
 <br />Pygments==2.15.1
 <br />pyOpenSSL @ file:///opt/conda/conda-bld/pyopenssl_1643788558760/work
 <br />pyparsing==3.1.2
 <br />pyproject_hooks==1.1.0
 <br />pyrsistent==0.19.3
 <br />PySocks @ file:///home/builder/ci_310/pysocks_1640793678128/work
 <br />pytest==8.2.2
 <br />pytest-sphinx==0.6.3
 <br />pytest-xdist==3.6.1
 <br />python-dateutil==2.8.2
 <br />python-dotenv==1.0.1
 <br />python-json-logger==2.0.7
 <br />pytz==2024.1
 <br />PyYAML==6.0
 <br />pyzmq==25.1.0
 <br />qtconsole==5.4.3
 <br />QtPy==2.3.1
 <br />ray==2.12.0
 <br />readme_renderer==43.0
 <br />referencing==0.35.0
 <br />regex==2024.4.28
 <br />requests @ file:///opt/conda/conda-bld/requests_1657734628632/work
 <br />requests-toolbelt==1.0.0
 <br />responses==0.18.0
 <br />rfc3339-validator==0.1.4
 <br />rfc3986==2.0.0
 <br />rfc3986-validator==0.1.1
 <br />rich==13.4.2
 <br />rpds-py==0.18.0
 <br />rsa==4.9
 <br />ruamel.yaml @ file:///croot/ruamel.yaml_1666304550667/work
 <br />ruamel.yaml.clib @ file:///croot/ruamel.yaml.clib_1666302247304/work
 <br />ruff==0.4.10
 <br />rusty-dawg @ file:///home/willm/rusty-dawg/bindings/python/target/wheels/rusty_dawg-0.1.0-cp310-cp310-manylinux_2_31_x86_64.whl#sha256=b780f6524d32a76e9dfff4137425f799401afc86deed08bbe1c73ac4885e3baf
 <br />s3transfer==0.10.1
 <br />safetensors==0.4.3
 <br />scipy==1.13.0
 <br />seaborn==0.13.2
 <br />SecretStorage==3.3.3
 <br />Send2Trash==1.8.2
 <br />sentencepiece==0.2.0
 <br />six @ file:///tmp/build/80754af9/six_1644875935023/work
 <br />smmap==5.0.0
 <br />sniffio==1.3.0
 <br />snowballstemmer==2.2.0
 <br />soupsieve==2.4.1
 <br />Sphinx==7.0.1
 <br />sphinx-autobuild==2021.3.14
 <br />sphinx-autodoc-typehints==1.23.3
 <br />sphinx-basic-ng==1.0.0b2
 <br />sphinx-copybutton==0.5.2
 <br />sphinxcontrib-applehelp==1.0.8
 <br />sphinxcontrib-devhelp==1.0.6
 <br />sphinxcontrib-htmlhelp==2.0.5
 <br />sphinxcontrib-jsmath==1.0.1
 <br />sphinxcontrib-qthelp==1.0.7
 <br />sphinxcontrib-serializinghtml==1.1.10
 <br />spinners==0.0.24
 <br />stack-data==0.6.2
 <br />starlette==0.37.2
 <br />sympy==1.12
 <br />tabulate==0.9.0
 <br />termcolor==2.4.0
 <br />terminado==0.17.1
 <br />tiktoken==0.6.0
 <br />tinycss2==1.2.1
 <br />tokenizers==0.19.1
 <br />tomli==2.0.1
 <br />toolz @ file:///croot/toolz_1667464077321/work
 <br />torch==2.2.2
 <br />tornado==6.3.2
 <br />tqdm @ file:///opt/conda/conda-bld/tqdm_1664392687731/work
 <br />traitlets==5.9.0
 <br />transformers==4.40.1
 <br />triton==2.2.0
 <br />twine==5.1.0
 <br />typeguard==2.13.3
 <br />typing_extensions==4.11.0
 <br />tzdata==2024.1
 <br />tzlocal==5.2
 <br />uri-template==1.2.0
 <br />urllib3 @ file:///croot/urllib3_1673575502006/work
 <br />uvicorn==0.29.0
 <br />uvloop==0.19.0
 <br />vllm==0.4.2
 <br />vllm-nccl-cu12==2.18.1.0.4.0
 <br />watchfiles==0.21.0
 <br />wcwidth==0.2.6
 <br />webcolors==1.13
 <br />webencodings==0.5.1
 <br />websocket-client==1.5.3
 <br />websockets==12.0
 <br />Werkzeug==3.0.3
 <br />widgetsnbextension==4.0.7
 <br />xformers==0.0.26.post1
 <br />xxhash==3.4.1
 <br />yarl==1.9.4
 <br />zipp==3.19.2
 <br />zstandard @ file:///opt/conda/conda-bld/zstandard_1663827383994/work
 <br />
</verbatim>
</details>

### PR DESCRIPTION
**Pull request by 2015aroras**
Issue: `Tokenizer.from_checkpoint` assumes that tokenizers are in HF or in a path relative to the current directory. This assumption doesn't hold when OLMo is install from pip as `ai2-olmo`. More generally, we don't have a clear mechanism for putting data files in our repo.

Fix: This PR creates an `olmo_data` package, of which the subdirectories can correspond to various types of data (e.g. `tokenizers` and `hf_datasets`). Tokenizers are moved under `olmo_data`, and `Tokenizer.from_checkpoint` is updated to look at local paths, then `olmo_data`, then HF Hub.

This change sets up the foundations for adding HF datasets to our repo, so that we don't have to make network calls during training runs.

Fixes #633

### CODE REVIEW

File: olmo_data/data.py [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+def is_data_dir(data_rel_path: str) -> bool:
+    return _get_data_traversable(data_rel_path).is_dir()
+
+
+def is_data_file(data_rel_path: str) -> bool:
+    return _get_data_traversable(data_rel_path).is_file()
+
+
+@contextmanager
+def get_data_path(data_rel_path: str) -> Iterator[Path]:
```
Comment (epwalsh) [inline olmo_data/data.py:22]: I think the correct return type is:
```suggestion
def get_data_path(data_rel_path: str) -> Generator[Path, None, None]:
```

**Review by epwalsh — APPROVED** [review]
LGTM, only one minor comment

File: olmo_data/data.py (hunk 1 above)
Comment (2015aroras) [reply olmo_data/data.py:22]: The [python docs](https://docs.python.org/3/library/typing.html#typing.Generator) say

> If your generator will only yield values, set the SendType and ReturnType to None:
> ...
> Alternatively, annotate your generator as having a return type of either Iterable[YieldType] or Iterator[YieldType]:

🤷

Comment (2015aroras) [reply olmo_data/data.py]: https://github.com/allenai/OLMo/pull/645/commits/d005b165a2e4c66bda8749bf9e166edbeddf7a08

Comment (epwalsh) [reply olmo_data/data.py]: Ok fair enough!

Comment (epwalsh) [reply olmo_data/data.py]: I'm with it either way. Your way is less verbose so that's a plus.

**Comment by OyvindTafjord** [comment]
@2015aroras Could we add something which doesn't break `scripts/convert_olmo_to_hf_new.py` for old configs? (i.e., [this stuff](https://github.com/allenai/OLMo/blob/1b7d2756326c465d986b15ca7c3011c5a1b0f1e1/scripts/convert_olmo_to_hf_new.py#L207))

**Pull request merged**